org.adempiere.base/src/org/compiere/model/MInvoice.java

```diff
@@ -1943,7 +1943,42 @@ else if (!isSOTrx()
 		if (matchPO > 0)
 			info.append(" @M_MatchPO_ID@#").append(matchPO).append(" ");
 
+		// Generate and complete an allocation if is a credit memo linked to an invoice
+		if (isCreditMemo() && getRelatedInvoice_ID() > 0) {
+			MInvoice inv = new MInvoice(getCtx(), getRelatedInvoice_ID(), get_TrxName());
 
+			if (!inv.isPaid()) {
+				BigDecimal invOpenAmt = inv.getOpenAmt();
+				BigDecimal cmOpenAmt = getOpenAmt();
+				StringBuilder aDescription = new StringBuilder(Msg.getElement(getCtx(), "C_Invoice_ID")).append(": ").append(inv.getDocumentNo()).append("/").append(getDocumentNo());
+
+				MAllocationHdr ah = new MAllocationHdr(getCtx(), true, TimeUtil.max(getDateAcct(), inv.getDateAcct()), getC_Currency_ID(), aDescription.toString(), get_TrxName());
+				ah.setAD_Org_ID(getAD_Org_ID());
+				ah.saveEx();
+
+				if (invOpenAmt.abs().compareTo(cmOpenAmt.abs()) == 0) {

```

Comment:
The usage of .abs() is mostly a bad practice for these kind of comparisons.

Problem here is that an invoice of +10 and a credit memo of +10 or -10 (a negative credit memo) - will compare as equal.

I think the comparison must be exact, not with .abs() - and same goes for the other two comparisons below - which I think it must take into account if is sales or purchase.